### PR TITLE
ID ejecting from computers by alt-clicking

### PR DESCRIPTION
**Pull request by kingofkosmos**
:cl:
add: Alt-clicking on a computer now ejects the ID card inside it.
/:cl:

What it says on the title.
Also makes the computer screen always refresh, and make the terminal_insert_disk sound, when you insert an ID.

### CODE REVIEW

**Comment by praisenarsie** [comment]
@kingofkosmos you misspelt something in the changelog

**Comment by kingofkosmos** [comment]
Whoops! Thanks.

**Comment by Qbopper** [comment]
god tier pr tbh

**Comment by epochayur** [comment]
god bless QoL PRs like this

File: code/game/machinery/computer/card.dm [hunk 1]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+			return
+		if(scan)
+			eject_id_scan()
+		else if(modify)
+			eject_id_modify()
+
+/obj/machinery/computer/card/proc/eject_id_scan(mob/user)
+	if(scan)
+		scan.forceMove(drop_location())
+		scan.verb_pickup()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user.put_in_hands(scan)

File: code/game/machinery/computer/card.dm [hunk 2]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+		playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+		scan = null
+	else //switching the ID with the one you're holding
+		var/obj/item/I = usr.get_active_held_item()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 3]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+		if(istype(I, /obj/item/card/id))
+			if(!usr.transferItemToLoc(I,src))
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 4]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+				return
+			playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+			scan = I
+	authenticated = 0
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: FALSE

File: code/game/machinery/computer/card.dm [hunk 5]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+	updateUsrDialog()
+
+/obj/machinery/computer/card/proc/eject_id_modify(mob/user)
+	if(modify)
+		GLOB.data_core.manifest_modify(modify.registered_name, modify.assignment)
+		modify.update_label()
+		modify.forceMove(drop_location())
+		modify.verb_pickup()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user.put_in_hands(modify)

File: code/game/machinery/computer/card.dm [hunk 6]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+		playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+		modify = null
+		region_access = null
+		head_subordinates = null
+	else //switching the ID with the one you're holding
+		var/obj/item/I = usr.get_active_held_item()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 7]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+		if(istype(I, /obj/item/card/id))
+			if (!usr.transferItemToLoc(I,src))
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 8]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
+				return
+			playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+			modify = I
+	authenticated = 0
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: FALSE

File: code/game/machinery/computer/security.dm [hunk 9]
```diff
@@ -790,3 +789,20 @@ What a mess.*/
 					if(!record2 || record2 == active2)
 						return 1
 	return 0
+
+/obj/machinery/computer/secure_data/AltClick(mob/user)
+	if(user.canUseTopic(src))
+		eject_id()
```
Comment (Cyberboss) [inline code/game/machinery/computer/security.dm]: user

File: code/game/machinery/computer/card.dm [hunk 10]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
 	updateUsrDialog()
 	return
 
+/obj/machinery/computer/card/AltClick(mob/user)
+	if(user.canUseTopic(src))
+		if(user.stat)
+			return
+		if(scan)
+			eject_id_scan()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 11]
```diff
@@ -550,6 +527,51 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
 	return
 
+/obj/machinery/computer/card/AltClick(mob/user)
+	if(user.canUseTopic(src))
+		if(user.stat)
+			return
+		if(scan)
+			eject_id_scan()
+		else if(modify)
+			eject_id_modify()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: user

File: code/game/machinery/computer/card.dm [hunk 12]
```diff
@@ -349,38 +355,9 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
-				scan = null
-			else
-				var/obj/item/I = usr.get_active_held_item()
-				if (istype(I, /obj/item/card/id))
-					if (!usr.transferItemToLoc(I,src))
-						return
-					playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
-					scan = I
-			authenticated = 0
+			eject_id_scan()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: usr

File: code/game/machinery/computer/card.dm [hunk 13]
```diff
@@ -349,38 +355,9 @@ GLOBAL_VAR_INIT(time_last_changed_position, 0)
...
-			else
-				var/obj/item/I = usr.get_active_held_item()
-				if (istype(I, /obj/item/card/id))
-					if (!usr.transferItemToLoc(I,src))
-						return
-					playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
-					modify = I
-			authenticated = 0
-
+			eject_id_modify()
```
Comment (Cyberboss) [inline code/game/machinery/computer/card.dm]: usr

File: code/game/machinery/computer/security.dm [hunk 14]
```diff
@@ -301,15 +308,7 @@ What a mess.*/
...
-				if(scan)
-					usr.put_in_hands(scan)
-					scan = null
-				else
-					var/obj/item/I = usr.is_holding_item_of_type(/obj/item/card/id)
-					if(I)
-						if(!usr.transferItemToLoc(I, src))
-							return
-						scan = I
+				eject_id()
```
Comment (Cyberboss) [inline code/game/machinery/computer/security.dm]: usr

File: code/game/machinery/computer/security.dm [hunk 15]
```diff
@@ -790,3 +789,20 @@ What a mess.*/
...
 						return 1
 	return 0
+
+/obj/machinery/computer/secure_data/AltClick(mob/user)
+	if(user.canUseTopic(src))
+		eject_id()
+
+/obj/machinery/computer/secure_data/proc/eject_id(mob/user)
+	if(scan)
+		usr.put_in_hands(scan)
```
Comment (Cyberboss) [inline code/game/machinery/computer/security.dm]: user

File: code/game/machinery/computer/security.dm [hunk 16]
```diff
@@ -790,3 +789,20 @@ What a mess.*/
...
+		scan = null
+		playsound(src, 'sound/machines/terminal_insert_disc.ogg', 50, 0)
+	else //switching the ID with the one you're holding
+		var/obj/item/I = usr.is_holding_item_of_type(/obj/item/card/id)
```
Comment (Cyberboss) [inline code/game/machinery/computer/security.dm]: user

File: code/game/machinery/computer/security.dm [hunk 17]
```diff
@@ -790,3 +789,20 @@ What a mess.*/
...
+		if(I)
+			if(!usr.transferItemToLoc(I, src))
```
Comment (Cyberboss) [inline code/game/machinery/computer/security.dm]: user

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
As a note to all, this won't work on modular computer id programs

**Comment by AnturK** [comment]
ID Scanner component would be neater.

**Comment by Cyberboss** [comment]
Modular computers already have their own sort of component kajiggery though

**Comment by kingofkosmos** [comment]
>this won't work on modular computer id programs

What do you mean? Modular computers already have alt-click-ejecting.

**Comment by Cyberboss** [comment]
Oh I forgot

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**